[CoreText] Measuring the width of multiple characters in one API call

brentfulgham opened the pull request:
#### 1b7efd59e1c0fc9df54b64bfdb5e34db9f99e78d
<pre>
[CoreText] Measuring the width of multiple characters in one API call
<a href="https://bugs.webkit.org/show_bug.cgi?id=289642">https://bugs.webkit.org/show_bug.cgi?id=289642</a>
<a href="https://rdar.apple.com/146888513">rdar://146888513</a>

Reviewed by NOBODY (OOPS!).

Instead of calling `CTFontGetAdvancesForGlyphs` on each character, pass a vector of characters
so we only pay to create the graphics context and select a font into it one time (instead of
once for each character).

This patch also tidies up some vector initialization to avoid reallocations, and removes
some unneeded `protect` calls on smart pointers already owned by the current object to
reduce refcount churn.

Profiling indicated that the batch width processing can introduce overhead for sufficiently
short text runs, so we use the single-character logic for runs up to 16 characters, but
batch for longer runs.

* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::processCharacterSpan): Added.
(WebCore::FontCascade::widthForSimpleTextSlow const): Use new multi-character measurement
when appropriate.
* Source/WebCore/platform/graphics/FontInlines.h:
(WebCore::Font::boundsForGlyphs const): Switch from size_t to uint32_t for glyph position
in the span (positionsNeedingMeasurement), which ranges between 0 and 255, and even if
we exceed that range we would still be able to handle a 4.3 billion character span with
the smaller size.
(WebCore::Font::widthForGlyph const):
(WebCore::Font::widthsForGlyphs const): Ditto.
(WebCore::Font::glyphsAndWidthsForCharacters const): Added. Handles glyph and width logic
in the same pass, rather than looping through the characters multiple times.
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::reserveInitialCapacity):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformWidthsForGlyphs const): Added.
(WebCore::Font::platformBoundsForGlyphs const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b7efd59e1c0fc9df54b64bfdb5e34db9f99e78d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/159304 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/32732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/25837 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/168134 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/113682 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/e01281ee-7d82-4bef-9cc4-3cdd8d4430c1) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/161173 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/32800 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/32719 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/123429 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/86643 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/7502a1bb-8f74-4eca-b2a3-cc0b1cab272a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/162261 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/25689 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/143107 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/104096 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/759282ef-d727-4aa1-ba10-4b4e94fd90c8) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/24742 "Found 1 new test failure: imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-window-stop.html (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/23183 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/15907 "Built successfully") | | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/134429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/20887 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/170628 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/16662 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/22513 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/131629 "Passed tests") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/32421 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/27255 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/131742 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/32365 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/142680 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/90490 "Built successfully") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/26425 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/19489 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/31876 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/98328 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/31396 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/31669 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/31551 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -366,21 +366,47 @@ float FontCascade::width(CodePath codePathToUse, const TextRun& run, SingleThrea
+#else
         for (size_t i = 0; i < characters.size(); ++i) {
             auto glyph = font.glyphForCharacter(characters[i]);
             glyphBuffer.add(glyph, font, font.widthForGlyph(glyph), i);
```
I'm confused, shouldn't `processCharacterSpan` be replacing something? We still call font.widthForGlyph for individual glyphs it seems.

brentfulgham replied:
The single-glyph widthForGlyph call is the intentional fast path for 1-character strings. The batch path (glyphsAndWidthsForCharacters) replaces the old multi-character flow.

vitorroriz replied:
oh sorry, I think my brain read #endif instead of else

brentfulgham committed:
[CoreText] Measuring the width of multiple characters in one API call
https://bugs.webkit.org/show_bug.cgi?id=289642
rdar://146888513

Reviewed by NOBODY (OOPS!).

Instead of calling `CTFontGetAdvancesForGlyphs` on each character, pass a vector of characters
so we only pay to create the graphics context and select a font into it one time (instead of
once for each character).

This patch also tidies up some vector initialization to avoid reallocations, and removes
some unneeded `protect` calls on smart pointers already owned by the current object to
reduce refcount churn.

Profiling indicated that the batch width processing can introduce overhead for sufficiently
short text runs, so we use the single-character logic for runs up to 16 characters, but
batch for longer runs.

* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::processCharacterSpan): Added.
(WebCore::FontCascade::widthForSimpleTextSlow const): Use new multi-character measurement
when appropriate.
* Source/WebCore/platform/graphics/FontInlines.h:
(WebCore::Font::boundsForGlyphs const): Switch from size_t to uint32_t for glyph position
in the span (positionsNeedingMeasurement), which ranges between 0 and 255, and even if
we exceed that range we would still be able to handle a 4.3 billion character span with
the smaller size.
(WebCore::Font::widthForGlyph const):
(WebCore::Font::widthsForGlyphs const): Ditto.
(WebCore::Font::glyphsAndWidthsForCharacters const): Added. Handles glyph and width logic
in the same pass, rather than looping through the characters multiple times.
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::reserveInitialCapacity):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformWidthsForGlyphs const): Added.
(WebCore::Font::platformBoundsForGlyphs const):
